[tapsend]: Enforce unique script keys

guggero opened the pull request:
Fixes https://github.com/lightninglabs/taproot-assets/issues/1168 (or at least prevents it from happening in the future).

This PR enforces script keys to be unique per top-level MS-SMT tree (assetID/groupKey) within a single transfer.

To be able to test this properly, we also make sure selected UTXOs are released by the freighter if the shipment of a parcel results in an error.
That change then impacts some assumptions in the integration tests, which requires additional commits to fix.

guggero commented:
I'll look into the failed itest on Monday, looks like a test is using duplicate keys.

Roasbeef commented on tapsend/send.go:
```diff
@@ -130,6 +130,14 @@ var (
 	ErrNoRootLocator = errors.New(
 		"cannot create split commitment without split root output",
 	)
+
```
In commit message: collidingon -> colliding

Roasbeef commented on tapsend/send.go:
```diff
@@ -364,6 +372,46 @@ func ValidateInputs(inputCommitments tappsbt.InputCommitments,
+		for idx := range vPkt.Outputs {
+			vOut := vPkt.Outputs[idx]
+
+			// Check if the script key has already been used in this
```
Isn't it ok for us to have a script key duplicate _across_ Bitcoin transaction outputs?

jharveyb replied:
I think a duplicate script key for the same asset would be an issue, even without those assets being children of an asset split. If that occurred, the exclusion proof would be a proof of inclusion, but for a _different_ leaf vs. the existing model where the exclusion proof terminates in a nil leaf.

Roasbeef replied:
Hmm yeh, I do wonder if there's a small change we can make the the exclusion proofs here to patch this portion.

Roasbeef commented on tapsend/send_test.go:
```diff
@@ -2952,3 +2952,151 @@ func TestValidateAnchorInputs(t *testing.T) {
+		vPkt.SetInputAsset(0, &a)
+
+		for i, outputKey := range outputKeys {
+			vPkt.Outputs[i] = &tappsbt.VOutput{
```
Shouldn't the output index be considered here when determining uniqueness?

guggero replied:
No, that's kind of the point of this fix. See comment(s) above.

jharveyb commented on tapsend/send.go:
```diff
@@ -364,6 +372,46 @@ func ValidateInputs(inputCommitments tappsbt.InputCommitments,
+// same script key in the same anchor output) or with exclusion proofs (e.g. two
+// asset outputs with the same script key in different anchor outputs).
+func ValidateScriptKeysUnique(packets []*tappsbt.VPacket) error {
+	// We'll keep track of all script keys (per top commitment key) that
```
:100:

jharveyb commented on tapsend/send_test.go:
```diff
@@ -2952,3 +2952,151 @@ func TestValidateAnchorInputs(t *testing.T) {
+			},
+		},
+		{
+			name: "collision, same group packets, same keys",
```
would be good to test collision based on asset ID + script key in addition to group key + script key.

guggero replied:
Thanks, added two more test cases.

jharveyb commented on tapfreighter/coin_select.go:
```diff
@@ -67,8 +67,14 @@ func (s *CoinSelect) SelectCoins(ctx context.Context,
+			return c.AnchorPoint.String()
+		},
+	)
+	log.Infof("Identified %v eligible asset inputs for send of %d to %v: "+
```
nit: this is counting anchor inputs and not asset inputs?

IIUC the anchor inputs may contain multiple eligible asset inputs, which we will select from later.

guggero replied:
I think if we have multiple eligible assets in the same anchor, we'd still get multiple `*AnchoredCommitment` structs returned. So IMO this is correct.

jharveyb commented on itest/send_test.go:
```diff
@@ -102,6 +103,16 @@ func testBasicSendUnidirectional(t *harnessTest) {
 
+	// Before we start sending, we test that we aren't allowed to send to
+	// the same address more than once within the same transfer.
+	_, err = t.tapd.SendAsset(ctxb, &taprpc.SendAssetRequest{
```
👍🏽

jharveyb reviewed (approved):
Looks solid, thanks for catching this!

We may want to add similar checks earlier in the freighter eventually but this should address the root cause IMO.

gijswijs reviewed (approved):
Very nice!

I feel that the tapfreighter commits should be their own PR? They don't have anything to do with enforcing unique script keys. Same thing goes fee estimation itest.

Or maybe just mention it in the PR that a number of other fixes were implemented as well, just for future reference.

gijswijs commented on tapsend/send.go:
```diff
@@ -364,6 +372,46 @@ func ValidateInputs(inputCommitments tappsbt.InputCommitments,
+			// Check if the script key has already been used in this
+			// transaction.
+			scriptKey := asset.ToSerialized(vOut.ScriptKey.PubKey)
+			if _, ok := perAssetMap[scriptKey]; ok {
```
Let me see if I understand this:
In theory it would be ok to reuse scriptKeys over different tapCommitmentKeys?

jharveyb replied:
Yes, since those would have different paths in the anchored `tapCommitment`.

I think this statement is correct:

"For a given transfer, across all anchor outputs, assets can share a `TapCommitmentKey()`, or an `AssetCommitmentKey()`, but not both."

If both are shared, for any two assets, then we can't generate valid exclusion proofs.

guggero commented:
> They don't have anything to do with enforcing unique script keys. Same thing goes fee estimation itest.

Actually, they kind of do. To be able to test the check introduced with the PR, we want to be able to try again. That's why we added the fix for releasing the leased inputs. Which then changes a bunch of assumptions in the itests.
Updated the PR body and commit messages to make that more clear.

guggero commented:
Thanks to the failing integration test I noticed two things that I needed to fix (with the latest push):
 - For grouped assets, the asset ID goes into the asset-level commitment key, not just the script key. So the commitment keys actually look like this (and uniqueness needs to be guaranteed across those within a single transfer):

|                   | TAP level           | asset level                        |
|-------------------|---------------------|------------------------------------|
| grouped asset     | `SHA256(group_key)` | `SHA256(asset_id \|\| script_key)` |
| non-grouped asset | `asset_id`          | `SHA256(script_key)`               |

 - When spending multiple input assets from the same on-chain input where other passive assets were present, we would duplicate those passive assets when creating passive virtual packets for them. This wasn't really an issue before since within the commitment trees they'd just be collapsed/overwritten into a single asset. But because the new uniqueness test operates on the slice of virtual packets, this was discovered. So a simple de-duplication fix was added in the last commit.

jharveyb commented:
>     * For grouped assets, the asset ID goes into the asset-level commitment key, not just the script key. So the commitment keys actually look like this (and uniqueness needs to be guaranteed across those within a single transfer):
> 
> 
> 	TAP level 	asset level
> grouped asset 	`SHA256(group_key)` 	`SHA256(asset_id || script_key)`
> non-grouped asset 	`asset_id` 	`SHA256(script_key)`

May be a motivation to use the `Specifier` type in more places.

guggero commented:
@jharveyb with the plan for getting rid of the collisions, I think we should go ahead and merge this. Would appreciate a review, thanks in advance!

jharveyb reviewed (approved):
Re-reviewed, LGTM 👍🏽
